### ISSUE TITLE
lint: always assign err before testing

### ISSUE THREAD

**Issue by quite** [go-critic/go-critic#837]
I've been told, and tend to agree that the following

```    
if os.Mkdir("some", 0755) != nil {
    t.Fatalf("couldn't mkdir "some")
}
```

should rather be written as the (arguably) more idiomatic:

```    
if err := os.Mkdir("some", 0755); err != nil {
    t.Fatalf("couldn't mkdir "some")
}
```
And I thought of go-critic. Does anybody agree? I find it more easy to read -- "OK, this if is about testing the err from Mkdir, the test is for non-nil".

**Comment by cristaloleg** [comment]
To be honest - I like it. I had the same idea regarding C/C++/others when you can do something like `if (foo) {...}` and `foo` is an int, pointer, anything.

**Comment by quite** [comment]
You're saying that you like the `if err := foo(); …` idiom, right? :)

**Comment by cristaloleg** [comment]
Exactly 😉

**Comment by quasilyte** [comment]
Can be implemented as a AST-based rule:
https://github.com/quasilyte/go-ruleguard/blob/70752dd040a0c4379b045bb854ce3dd7b0912d5c/analyzer/testdata/src/extra/extra.rules.go#L26-L29

```go
m.Match(`$err != nil`,
	`$err == nil`).
	Where(!m["err"].Pure && m["err"].Type.Is(`error`)).
	Report(`assign $err to err and then do a nil check`)
```

**Comment by quite** [comment]
That's neat! go-critic is not currently using go-ruleguard, right?

**Comment by quasilyte** [comment]
It's not, but when it will be more stable, it probably could.
